multicluster/controllers/multicluster/clusterset_controller.go

```diff
@@ -242,43 +242,24 @@ func (r *ClusterSetReconciler) updateMultiClusterSetOnMemberCluster(clusterSet *
 
 	klog.InfoS("ClusterSet update", "addedLeaders", addedLeaders, "removedLeaders", removedLeaders)
 
-	ch := make(chan error)
-	var wg sync.WaitGroup
-	wg.Add(len(addedLeaders))
-	go func() {
-		wg.Wait()
-		close(ch)
-	}()
-
+	var multiErr error
 	for _, addedLeader := range addedLeaders {
 		clusterID := common.ClusterID(addedLeader.ClusterID)
 		url := addedLeader.Server
 		secretName := addedLeader.Secret
 
 		klog.InfoS("Creating RemoteCluster", "clusterID", clusterID)
-
-		go func(clusterID common.ClusterID, url string, secretName string) {
-			defer wg.Done()
-
-			_, err := core.NewRemoteCluster(clusterID, r.clusterSetID, url, secretName, r.Scheme,
-				r.LocalClusterManager, r.RemoteClusterManager, clusterSet.Spec.Namespace,
-				clusterSet.GetNamespace())
-			if err != nil {
-				klog.ErrorS(err, "Unable to create RemoteCluster", "clusterID", clusterID)
-			} else {
-				klog.InfoS("Created RemoteCluster", "clusterID", clusterID)
-			}
-
-			ch <- err
-		}(clusterID, url, secretName)
-	}
-
-	var err error
-	for errFromCh := range ch {
-		if errFromCh != nil {
-			klog.ErrorS(errFromCh, "Received error")
-			err = multierr.Append(err, errFromCh)
+		// Read secret to access the leader cluster. Assume secret is present in the same Namespace as the ClusterSet.
+		secret := r.getSecretForLeader(secretName, clusterSet.GetNamespace())
+		_, err := core.NewRemoteCluster(clusterID, r.clusterSetID, url, secret, r.Scheme,
+			r.LocalClusterManager, r.RemoteClusterManager, clusterSet.Spec.Namespace,
+			clusterSet.GetNamespace())
+		if err != nil {
+			klog.ErrorS(err, "Unable to create RemoteCluster", "clusterID", clusterID)
```

Comment:
Question - what will happen for failure of one remote cluster? Do we retry later?

Reply:
Yes we aggregate the errors from here and return it, so reconciliation is considered failed and will be retried.